polar-core/src/vm.rs

```diff
@@ -758,37 +795,46 @@ impl PolarVirtualMachine {
     }
 
     fn log(&self, message: &str, terms: &[&Term]) {
-        self.log_with(|| message, terms)
+        self.log_with(LogLevel::Info, || message, terms)
     }
 
-    fn log_with<F, R>(&self, message_fn: F, terms: &[&Term])
+    fn log_with<F, R>(&self, level: LogLevel, message_fn: F, terms: &[&Term])
     where
         F: FnOnce() -> R,
         R: AsRef<str>,
     {
-        if self.polar_log && !self.polar_log_mute {
-            let mut indent = String::new();
-            for _ in 0..=self.queries.len() {
-                indent.push_str("  ");
-            }
-            let message = message_fn();
-            let lines = message.as_ref().split('\n').collect::<Vec<&str>>();
-            if let Some(line) = lines.first() {
-                let mut msg = format!("[debug] {}{}", &indent, line);
-                if !terms.is_empty() {
-                    let relevant_bindings = self.relevant_bindings(terms);
-                    msg.push_str(&format!(
-                        ", BINDINGS: {{{}}}",
-                        relevant_bindings
-                            .iter()
-                            .map(|(var, val)| format!("{} => {}", var.0, val.to_polar()))
-                            .collect::<Vec<String>>()
-                            .join(", ")
-                    ));
+        if let Some(configured_log_level) = self.log_level {
+            // preserve the old `polar_log_mute` behavior which omits parameter
+            // specialization checking Unify, IsA and other events from the log
+            if level == LogLevel::Trace && self.polar_trace_mute {
```

Comment:
This feels odd to me — `Trace` is supposed to be the most verbose level, so why would we ever want to skip logs for `Trace` but not for the other levels?

Reply:
we currently temporarily mute the `POLAR_LOG=trace` output for a period of the VM execution starting before rule filtering and resuming again in `sort_rules` once we have filtered & sorted all applicable rules. I understand from @saolsen that this is intended to prevent the log from being cluttered with these parameter related outputs.
https://github.com/osohq/oso/blob/main/polar-core/src/vm.rs#L1520
https://github.com/osohq/oso/blob/main/polar-core/src/vm.rs#L2582

Reply:
Right, I understand why we want to early return if `self.polar_trace_mute` is `true`. I'm wondering whether we need the `level == LogLevel::Trace` condition here. Maybe the answer is that currently we'll only ever encounter `log_with(LogLevel::Trace, ...)` calls while the `polar_trace_mute` bit is flipped, but maybe that won't always be the case.

Have you done any thinking on whether any sort of non-trace logging would be useful during rule selection and sorting? I guess that's kind of related to the "why a rule was filtered out" logic that's coming up soon. What are your thoughts?

Reply:
I believe you are correct that we only currently encounter `log_with(LogLevel::Trace)`, however I can definitely see us expanding the set of logs to include "why a rule didn't match" style messages in the not too distant future. My thinking on that was to use the DEBUG level for these logs, in which case we'd want to retain the `level == LogLevel::Trace` condition and allow them to pass.